File: src/main/java/org/oewntk/pojos/Pos.java
/*
 * Copyright (c) 2021. Bernard Bou.
 */

package org.oewntk.pojos;

/**
 * Pos (part of speech)
 *
 * @author Bernard Bou
 */
public enum Pos
{
	NOUN('n', "noun", "noun"), //
	VERB('v', "verb", "verb"), //
	ADJ('a', "adj", "adjective"), //
	ADV('r', "adv", "adverb");

	private final char id;

	private final String name;

	private final String description;

	/**
	 * Constructor
	 *
	 * @param id          character id
	 * @param name        name
	 * @param description description
	 */
	Pos(final char id, final String name, final String description)
	{
		assert id == 'n' || id == 'v' || id == 'a' || id == 'r';
		this.id = id;
		this.name = name;
		this.description = description;
	}

	/**
	 * Get char id
	 *
	 * @return char id
	 */
	public char toChar()
	{
		return this.id;
	}

	/**
	 * Get name
	 *
	 * @return name
	 */
	public String getName()
	{
		return this.name;
	}

	/**
	 * Get description
	 *
	 * @return description
	 */
	public String getDescription()
	{
		return this.description;
	}

	/**
	 * Parse pos from character id
	 *
	 * @param id character id
	 * @return pos
	 * @throws ParsePojoException parse exception
	 */
	public static Pos parsePos(final char id) throws ParsePojoException
	{
		for (final Pos pos : Pos.values())
		{
			if (id == pos.id)
			{
				return pos;
			}
		}
		throw new ParsePojoException("Pos:" + id);
	}

	/**
	 * Make pos from pos index
	 *
	 * @param index0 index
	 * @return pos
	 */
	public static Pos fromIndex(final int index0)
	{
		final int index = index0 - 1;
		if (index >= 0 && index < Pos.values().length)
		{
			return Pos.values()[index];
		}
		throw new IllegalArgumentException("Pos:" + index);
	}

	/**
	 * Make pos from name
	 *
	 * @param name name
	 * @return pos
	 */
	public static Pos fromName(final String name)
	{
		for (final Pos pos : Pos.values())
		{
			if (name.equals(pos.name))
			{
				return pos;
			}
		}
		return null;
	}

	@Override
	public String toString()
	{
		return Character.toString(this.id);
	}
}

File: src/main/java/org/oewntk/pojos/Pos.kt
/*
 * Copyright (c) 2021-2024. Bernard Bou.
 */
package org.oewntk.pojos

/**
 * Pos (part of speech)
 *
 * @param id          character id
 * @param name2       name2
 * @param description description
 *
 * @author Bernard Bou
 */
enum class Pos(
    val id: Char,
    val name2: String,
    val description: String,
) {

    NOUN('n', "noun", "noun"),
    VERB('v', "verb", "verb"),
    ADJ('a', "adj", "adjective"),
    ADV('r', "adv", "adverb");

    init {
        assert(id == 'n' || id == 'v' || id == 'a' || id == 'r')
    }

    /**
     * Get char id
     *
     * @return char id
     */
    fun toChar(): Char {
        return this.id
    }

    override fun toString(): String {
        return id.toString()
    }

    companion object {

        /**
         * Parse pos from character id
         *
         * @param id character id
         * @return pos
         * @throws ParsePojoException parse exception
         */
        @Throws(ParsePojoException::class)
        fun parsePos(id: Char): Pos {
            return entries
                .firstOrNull { id == it.id } ?: throw ParsePojoException("Pos:$id")
        }

        /**
         * Make pos from name
         *
         * @param name name
         * @return pos
         */
        fun fromName(name: String): Pos {
            return entries
                .firstOrNull { name == it.name2 } ?: throw IllegalArgumentException(name)
        }
    }
}
